primitives/src/config/substrate_config.rs

```diff
@@ -0,0 +1,252 @@
+// Copyright 2019-2022 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+//! Substrate specific configuration
+//!
+//! This file is mostly subxt.
+//! https://github.com/paritytech/subxt/blob/ce0a82e3227efb0eae131f025da5f839d9623e15/subxt/src/config/substrate.rs
+
+use crate::{
+	config::{Config, Hasher, Header},
+	types::AccountData,
+	AssetTip, ExtrinsicSigner, GenericAdditionalParams, GenericExtrinsicParams,
+};
+use alloc::{format, vec::Vec};
+use codec::{Decode, Encode};
+use core::{fmt::Debug, marker::PhantomData};
+pub use primitive_types::{H256, U256};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use sp_core::sr25519;
+use sp_runtime::{traits::Block as BlockTrait, AccountId32, MultiAddress, MultiSignature};
+
+/// Default set of commonly used types by Substrate runtimes.
+#[derive(Decode, Encode, Clone, Eq, PartialEq, Debug)]
+pub struct SubstrateConfig<Block> {
+	_marker: PhantomData<Block>,
+}
+
+impl<Block> Config for SubstrateConfig<Block>
+where
+	Block: BlockTrait + DeserializeOwned,
+{
+	type Index = u32;
+	type Hash = H256;
+	type AccountId = AccountId32;
+	type Address = MultiAddress<Self::AccountId, u32>;
+	type Signature = MultiSignature;
+	type Hasher = BlakeTwo256;
+	type Header = SubstrateHeader<u32, BlakeTwo256>;
+	type AccountData = AccountData<Self::Balance>;
+	type ExtrinsicParams = SubstrateExtrinsicParams<Self>;
+	type CryptoKey = sr25519::Pair;
+	type ExtrinsicSigner = ExtrinsicSigner<Self, Self::CryptoKey>;
+	type Block = Block;
+	type Balance = u128;
+	type ContractBalance = u128;
+	type StakingBalance = u128;
+}
+
+/// A struct representing the signed extra and additional parameters required
+/// to construct a transaction for the default substrate node.
+pub type SubstrateExtrinsicParams<T> = GenericExtrinsicParams<T, AssetTip<u128>>;
+
+/// A builder which leads to [`SubstrateExtrinsicParams`] being constructed.
+/// This is what you provide to methods like `sign_and_submit()`.
+pub type SubstrateExtrinsicParamsBuilder<T> = GenericAdditionalParams<T, AssetTip<u128>>;
+
+/// A type that can hash values using the blaks2_256 algorithm.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Encode)]
+pub struct BlakeTwo256;
+
+impl Hasher for BlakeTwo256 {
```

Comment:
subxt inserted this in https://github.com/paritytech/subxt/commit/b316301d61f924c510c40f505995f28a83c1365f#diff-57540144ef1381f66537718f59e717110353e5408cdf539b67f7a1ff9cdd0d10 to make sp_core and sp_runtime dependencies optional. See https://github.com/paritytech/subxt/pull/760.
The Header trait has also been inserted for this purpose.
Removing these would certainly simplify our configurations and code. If our goal is to have an easy to use configuration, we should do it. But I like subxt idea of  "Reducing the need to keep up to date with Substrate packlage versions". See PR  https://github.com/paritytech/subxt/pull/760

Reply:
Since we're not using package versions anyway, that's not really a problem for us. However, I do see the point of wasm and huge dependency tree. But in that case a separate issue is necessary because we currently do depend on sp_core and sp_runtime. Not sure how easily removable that is.

If you think it worth keeping this trait, that's fine. But it probably makes sense to move the definition to another file for better visibility. We already have two dedicated files with substrate types:
- https://github.com/scs/substrate-api-client/blob/master/primitives/src/types.rs
- https://github.com/scs/substrate-api-client/blob/master/primitives/src/serde_impls.rs